_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,340 @@
+You just need an environment with few command line tools (curl, tar, java, mvn).
+
+This application consumes text messages from the `input-topic`, reverts their content, and sends the result to the `output-topic`.
+The following diagram shows the relevant RPCs.
+
+![txn app](/assets/images/posts/2023-04-14-kafka-txn-app.png)
+
+In the following snippet we start the cluster, run the application, and create a single transaction.
+For the sake of simplicity, we use a single-node Kafka cluster running on localhost, but the example also works with a multi-node cluster.
+
+```sh
+# download the Apache Kafka binary and start a single-node Kafka cluster in background
+$ KAFKA_VERSION="3.4.0" KAFKA_HOME="$(mktemp -d -t kafka.XXXXXXX)" && export KAFKA_HOME
+
+$ curl -sLk "https://archive.apache.org/dist/kafka/$KAFKA_VERSION/kafka_2.13-$KAFKA_VERSION.tgz" \
+  | tar xz -C "$KAFKA_HOME" --strip-components 1
+
+$ $KAFKA_HOME/bin/zookeeper-server-start.sh -daemon $KAFKA_HOME/config/zookeeper.properties \
+  && sleep 5 && $KAFKA_HOME/bin/kafka-server-start.sh -daemon $KAFKA_HOME/config/server.properties
+
+# open a new terminal, get and run the application (Ctrl+C to stop)
+$ git clone git@github.com:fvaleri/examples.git -n --depth=1 --filter=tree:0 && cd examples \
+  && git sparse-checkout set --no-cone kafka/kafka-txn && git checkout 5fe575a16fd4106519b13953f2049a6c3101e9f4
+...
+HEAD is now at 5fe575a Add invalid offset handling
+
+$ export BOOTSTRAP_SERVERS="localhost:9092" INSTANCE_ID="kafka-txn-0" \
+  GROUP_ID="my-group" INPUT_TOPIC="input-topic" OUTPUT_TOPIC="output-topic"
+
+$ mvn clean compile exec:java -q -f kafka/kafka-txn/pom.xml
+Starting instance kafka-txn-0
+Created topics: input-topic
+Waiting for new data
+...
+
+# come back to the previous terminal and send a test message
+$ $KAFKA_HOME/bin/kafka-console-producer.sh --bootstrap-server :9092 --topic input-topic
+>test
+>^C
+```
+
+All of this is pretty normal, but now comes the interesting part.
+
+How can we identify which partitions are involved in this transaction and inspect their content?
+We are writing to the `output-topic` which has only one partition, but the internal topics for storing consumer group offsets and transaction states have 50 partitions each by default.
+
+To avoid looking at 100 partitions in the worst case scenario, we can use the same hashing function that Kafka uses to find the coordinating partition.
+This function maps `group.id` to `__consumer_offsets` partition, and `transactional.id` to `__transaction_state` partition.
+
+```sh
+# define the function and find the coordinating partitions
+$ kafka-cp() {
+  local id="${1-}" part="${2-50}"
+  if [[ -z $id ]]; then echo "Missing id parameter" && return; fi
+  echo 'public void run(String id, int part) { System.out.println(abs(id.hashCode()) % part); }
+    private int abs(int n) { return (n == Integer.MIN_VALUE) ? 0 : Math.abs(n); }
+    run("'"$id"'", '"$part"');' | jshell -
+}
+
+$ kafka-cp my-group
+12
```

Comment:
Can we annotate this is a comment `# to the applications offsets are in partition 12 of __consumer_offsets`?